crates/core/tedge_watchdog/src/systemd_watchdog.rs

```diff
@@ -0,0 +1,142 @@
+    TEdgeConfigLocation,
+};
+
+#[derive(Serialize, Deserialize)]
+pub struct HealthStatus {
+    status: String,
+    pid: u32,
+}
+
+pub async fn start_watchdog(tedge_config_dir: PathBuf) -> Result<(), anyhow::Error> {
+    // Send ready notification to systemd.
+    notify_systemd(process::id(), "--ready")?;
+
+    let tedge_services = vec![
+        "tedge-mapper-c8y",
+        "tedge-mapper-az",
+        "tedge-mapper-collectd",
+        "tedge-agent",
+    ];
+
+    let watchdog_threads = FuturesUnordered::new();
+
+    for service in tedge_services {
+        let interval = match get_watchdog_sec(&format!("/lib/systemd/system/{service}.service")) {
+            Ok(i) => i,
+            Err(_e) => continue, // Watchdog not enabled for this service
+        };
+        let req_topic = format!("tedge/health-check/{service}");
+        let res_topic = format!("tedge/health/{service}");
+        let tedge_config_location =
+            tedge_config::TEdgeConfigLocation::from_custom_root(tedge_config_dir.clone());
+        if interval > 0 {
+            watchdog_threads.push(tokio::spawn(async move {
+                monitor_tedge_service(
+                    tedge_config_location,
+                    service,
+                    &req_topic,
+                    &res_topic,
+                    interval / 2,
+                )
+                .await
+            }));
+        }
+    }
+
+    futures::future::join_all(watchdog_threads).await;
+
+    Ok(())
+}
+
+async fn monitor_tedge_service(
+    tedge_config_location: TEdgeConfigLocation,
+    name: &str,
+    req_topic: &str,
+    res_topic: &str,
+    interval: u64,
+) -> Result<(), WatchdogError> {
+    let client_id: &str = &format!("{}_{}", name, nanoid!());
+    let mqtt_config = get_mqtt_config(tedge_config_location, client_id)?
+        .with_subscriptions(res_topic.try_into()?);
+    let client = mqtt_channel::Connection::new(&mqtt_config).await?;
+    let mut received = client.received;
+    let mut publisher = client.published;
+
+    println!("Starting watchdog for {} service", name);
+
+    loop {
+        let message = Message::new(&Topic::new(req_topic)?, "");
+        let _ = publisher
+            .publish(message)
+            .await
+            .map_err(|e| eprintln!("Publish failed with error: {}", e));
+
+        let start = Instant::now();
+
+        match tokio::time::timeout(tokio::time::Duration::from_secs(interval), received.next())
+            .await
+        {
+            Ok(Some(msg)) => {
+                let message = msg.payload_str()?;
+
+                let p: HealthStatus = serde_json::from_str(message)?;
+
+                notify_systemd(p.pid, "WATCHDOG=1")?;
+            }
+            Ok(None) => {}
+            Err(elapsed) => {
+                eprintln!("The {name} failed with {elapsed}");
+            }
+        }
+        tokio::time::sleep(start.elapsed()).await;
```

Comment:
Ok, now it looks better, but I think you forgot to subtract your elapsed from interval?

Something like:

```suggestion
        tokio::time::sleep(interval - start.elapsed()).await;
```

Reply:
Oops, yes missed it. Fixed now.